packages/tlsync/src/lib/squish.ts

```diff
@@ -0,0 +1,174 @@
+import { UnknownRecord } from '@tldraw/store'
+import { exhaustiveSwitchError, objectMapEntries, structuredClone } from '@tldraw/utils'
+import {
+	NetworkDiff,
+	ObjectDiff,
+	RecordOp,
+	RecordOpType,
+	ValueOpType,
+	applyObjectDiff,
+} from './diff'
+import { TLSocketServerSentDataEvent } from './protocol'
+
+interface State<R extends UnknownRecord> {
+	lastPatch: (TLSocketServerSentDataEvent<R> & { type: 'patch' }) | null
+	squished: TLSocketServerSentDataEvent<R>[]
+}
+
+type Bailed = boolean
+
+function patchThePatch(lastPatch: ObjectDiff, newPatch: ObjectDiff): Bailed {
+	for (const [newKey, newOp] of Object.entries(newPatch)) {
+		switch (newOp[0]) {
+			case ValueOpType.Put:
+				lastPatch[newKey] = newOp
+				break
+			case ValueOpType.Append:
+				if (lastPatch[newKey] === undefined) {
+					lastPatch[newKey] = newOp
+				} else {
+					const lastOp = lastPatch[newKey]
+					if (lastOp[0] === ValueOpType.Append) {
+						const lastValues = lastOp[1]
+						const lastOffset = lastOp[2]
+						const newValues = newOp[1]
+						const newOffset = newOp[2]
+						if (newOffset === lastOffset + lastValues.length) {
+							lastValues.push(...newValues)
+						} else {
+							// something weird is going on, bail out
+							return true
+						}
+					} else {
+						// bail out, it's too hard
+						return true
+					}
+				}
+				break
+			case ValueOpType.Patch:
+				if (lastPatch[newKey] === undefined) {
+					lastPatch[newKey] = newOp
+				} else {
+					// bail out, recursive patching is too hard
+					return true
+				}
+				break
+			case ValueOpType.Delete:
+				// overwrite whatever was there previously, no point if it's going to be removed
+				// todo: check if it was freshly put and don't add if it wasn't?
+				lastPatch[newKey] = newOp
+				break
+			default:
+				exhaustiveSwitchError(newOp[0])
+		}
+	}
+
+	return false
+}
+
+function patchTheOp<R extends UnknownRecord>(
+	lastRecordOp: RecordOp<R>,
+	newPatch: ObjectDiff
+): Bailed {
+	switch (lastRecordOp[0]) {
+		case RecordOpType.Put:
+			// patching a freshly added value is easy, just patch as normal
+			lastRecordOp[1] = applyObjectDiff(lastRecordOp[1], newPatch)
+			break
+		case RecordOpType.Patch: {
+			// both are patches, merge them
+			const bailed = patchThePatch(lastRecordOp[1], newPatch)
+			if (bailed) {
+				return true
+			}
+			break
+		}
+		case RecordOpType.Remove:
+			// we're trying to patch an object that was removed, just disregard the update
```

Comment:
this should never happen, might be worth adding an assert

Reply:
commented below re: assertions